demos/skeletons/provider/template_kem.c

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/crypto.h>
+#include <openssl/evp.h>
+#include <openssl/core_dispatch.h>
+#include <openssl/core_names.h>
+#include <openssl/params.h>
+#include <openssl/err.h>
+#include <openssl/proverr.h>
+#include "prov/provider_ctx.h"
+#include "prov/implementations.h"
+#include "prov/securitycheck.h"
+#include "prov/providercommon.h"
```

Comment:
fwiw, my initial intent was to use them to create the various PQ provider function sets we need, i.e., for internal use, so the placement indeed seems wrong if "demos" is only for external consumption. I originally had them in providers/implementations without "wiring" into a provider. There they also built nicely and I could move them back there.

That said, a second intent was indeed to leave this as a "quick starting point" for others to do providers, both internal as well as external as I think the hurdle to create providers right now is pretty high. This was meant to lower it -- but again, this looks like a separate effort...

Reply:
I'd suggest somewhere under `providers/implementations`.
Agreed, the starting point for third party developers is a different project IMO.

Reply:
> I'd suggest somewhere under providers/implementations

Good for me. If so agreed (?) the only open question is how to ascertain it's "test built" there: OK to add it to the "libdefault.a" target without actually registering the symbols in "defltprov.c"? Other proposals welcome.

Reply:
Not sure about the "test built" issue. Perhaps @levitte has an opinion?

Reply:
> Not sure about the "test built" issue.

fwiw, latest commit shows the proposal.

Reply:
> OK to add it to the "libdefault.a" target without actually registering the symbols in "defltprov.c"?

No, that will mean it will get into libcrypto.so AFAIK.

Reply:
> No, that will mean it will get into libcrypto.so AFAIK.

Confirmed: It does, even though the symbols are not used anywhere :( So shall we then create a dummy library which to build this for/into to ascertain the code builds OK (or not build it at all)?

Reply:
a dummy library works for me

Reply:
so implemented.

Reply:
ok to mark conversation resolved @mattcaswell ?